### PR TITLE
feat: deprecated AMIs should be discoverable when specifying ami id on the amiSelectorTerms

### LINKED ISSUE #6393: bug: deprecated AMIs should be able to be listed when specifying ami id on the amiSelectorTerms
### Description

**Observed Behavior**:
We pin amis using ami id on our ec2 node class to be able to turn on drift and avoid untested ami's from reaching production. With this we ran into an issue where an deprecated AMI that is not owned by our AWS account was not discovered by Karpeneter when creating a new nodeclaim from an existing nodepool/ec2nodeclass. 

An AMI that is deprecated is still discoverable and can be used https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ami-deprecate.html this is evident from using the aws cli (see the example below)

<details>
  <summary>aws cli describe-images</summary>
 
    ✗ aws ec2 describe-images --region us-west-2 --image-ids ami-002c17c43a2d082b5
    {
        "Images": [
            {
                "Architecture": "x86_64",
                "CreationDate": "2024-03-17T07:09:41.000Z",
                "ImageId": "ami-002c17c43a2d082b5",
                "ImageLocation": "*******794/***-amazon-eks-node-1.27-v20240307-yellow-MTcxMDY1NTcyMC0y",
                "ImageType": "machine",
                "Public": false,
                "OwnerId": "*******794",
                "PlatformDetails": "Linux/UNIX",
                "UsageOperation": "RunInstances",
                "State": "available",
                "BlockDeviceMappings": [
                    {
                        "DeviceName": "/dev/xvda",
                        "Ebs": {
                            "DeleteOnTermination": true,
                            "Iops": 3000,
                            "SnapshotId": "snap-0cfc13de6d2d8d426",
                            "VolumeSize": 10,
                            "VolumeType": "gp3",
                            "Throughput": 125,
                            "Encrypted": true
                        }
                    }
                ],
                "Description": "*** EKS Kubernetes Worker AMI with AmazonLinux2 image, (k8s: 1.27.9, containerd: 1.7.*)",
                "EnaSupport": true,
                "Hypervisor": "xen",
                "Name": "***-amazon-eks-node-1.27-v20240307-yellow-MTcxMDY1NTcyMC0y",
                "RootDeviceName": "/dev/xvda",
                "RootDeviceType": "ebs",
                "SriovNetSupport": "simple",
                "Tags": [ 
                 ....
                 ....
                 ....
                ],
                "VirtualizationType": "hvm",
                "DeprecationTime": "2024-06-19T06:09:00.000Z"
            }
        ]
    }

</details>

The above ami as can be seen on the `DeprecationTime` field was deprecated and Karpenter failed to find the ami when provisioning a new node for an existing nodepool/ec2nodeclass. 

When scale up event happened on the node pool we saw the following message 
```
no ami matching constraint found 
```

An example our ec2nodeclass

<details>
  <summary>ec2 node class</summary>

    ✗ k get ec2nodeclass sandbox -o yaml
    apiVersion: karpenter.k8s.aws/v1beta1
    kind: EC2NodeClass
    metadata:
      annotations:
        karpenter.k8s.aws/ec2nodeclass-hash: "2915896952981468854"
        kubernetes.io/description: NodePool provisioned for ASL sandbox in dev-cluster
        meta.helm.sh/release-name: sandbox-workers
        meta.helm.sh/release-namespace: sandbox
      creationTimestamp: "2024-04-04T16:56:25Z"
      finalizers:
      - karpenter.k8s.aws/termination
      generation: 2
      labels:
        app.kubernetes.io/managed-by: Helm
      name: sandbox
      resourceVersion: "981834955"
      uid: e17e714f-9b90-4119-aee8-225b66f6574f
    spec:
      amiFamily: AL2
      amiSelectorTerms:
      - id: ami-002c17c43a2d082b5
 
</details>
   

Digging further into the code we saw that there is a field on the sdk https://github.com/aws/aws-sdk-go-v2/blob/main/service/ec2/api_op_DescribeImages.go#L183 which will enable listing deprecated amis for when an ami is not owned by the current aws account. This is by default set to false but it will list deprecated ami which are owned by the current aws account itself but not when the owner is different. 

The amiFamily controller might need to add this flag when the owner of the ami is different than the current aws account id https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/amifamily/ami.go

> NOTE: These are private AMIs not owned by AWS

**Expected Behavior**:

Karpenter should be able to detect deprecated amis and not fail provisioning new node claims and instances

**Reproduction Steps** (Please include YAML):

1. Use an AMI that is deprecated and not owned by the AWS account where Karpenter is running.
2. Reference that AMI in the ec2nodeclass using `amiSelectorTerms` and `id` 
3. Wait for Karpenter to provision the nodeclaim and fail with an message on the status `no ami matching constraint found` 

**Versions**:
- Chart Version:
- Kubernetes Version (`kubectl version`):

```
✗ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.2", GitCommit:"fc04e732bb3e7198d2fa44efa5457c7c6f8c0f5b", GitTreeState:"clean", BuildDate:"2023-02-22T13:39:03Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.13-eks-3af4770", GitCommit:"4873544ec1ec7d3713084677caa6cf51f3b1ca6f", GitTreeState:"clean", BuildDate:"2024-04-30T03:31:44Z", GoVersion:"go1.21.9", Compiler:"gc", Platform:"linux/amd64"}
```

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by shabbskagalwala**
Fixes #6393 

**Description**
Deprecated AMI's should be discovered by Karpenter for existing node pool to avoid disruptions to workloads in production systems. This change includes a new field on the `ec2nodeclass` called `amiIncludeDeprecated` of type `bool`.

This field has some CEL validation with what it does is

1. Can only be specified with AMI ids. Any other attributes in the `amiSelectorTerms` will cause the validation to fail.
2. No `name` attribute can be provided with the field for the reason that `name` attribute can contain wildcard which enables node pool to select AMIs which are available. 
3. This field is a global field in the sdk as well and applies to all AMI ids in `amiSelectorTerms`, to reduce the number of describe image api calls.

**How was this change tested?**

This was tested using 
1. `make verify` - to update the CRD definitions and apply to a development EKS cluster
2. `KO_DOCKER_REPO=4xxxxxxxxxx5.dkr.ecr.us-west-2.amazonaws.com/shabbskagalwala/karpenter-provider-aws/controller make image` to build and push the image, which was consumed in EKS.
3. Tests have been added to webhook and cel for validations
4. Updating the `ec2nodeclass` to use a deprecated ami `ami-002c17c43a2d082b5` which resulted in Karpenter successfully creating a nodeclaim with no errors as seen below

Additionally, the CEL rules were tested using the `kubectl apply` on a test nodepool using the combination of following parameters

```yaml
  amiIncludeDeprecated: true
  amiSelectorTerms:
    - name: "amazon-eks-arm64-node-1.27-*"
      owner: "0xxxxxxxxxx4"
    - id: "ami-0d20e6af81d7ce566"

  amiSelectorTerms:
    - name: "amazon-eks-arm64-node-1.27-*"
      owner: "0xxxxxxxxxx4"

  amiSelectorTerms:
    - id: "ami-0d20e6af81d7ce566"
    - id: "ami-002c17c43a2d082b5"

  amiSelectorTerms:
  - tags:
      karpenter.sh/discovery: "dev-cluster"
      environment: test
  - id: "ami-0d20e6af81d7ce566"
 ```

Example
For a nodepool with the following 

``` yaml
  amiIncludeDeprecated: true
  amiSelectorTerms:
  - tags:
      karpenter.sh/discovery: "dev-cluster"
      environment: test
  - id: "ami-0d20e6af81d7ce566"
```
Would return
``` bash
✗ k apply -f nodepool-test.yaml --dry-run=server
nodepool.karpenter.sh/cosmic-lamb created (server dry run)
The EC2NodeClass "cosmic-lamb" is invalid: spec: Invalid value: "object": 'amiIncludeDeprecated' is mutually exclusive, can only be set with ids in amiSelectorTerms
```

Using with deprecated AMIs Karpenter was successfully able to create nodeclaims for existing nodepools

<details>
  <summary>nodeclaim using deprecated ami</summary>

    ✗ k describe nodeclaim cosmic-lamb-jwjwm
    Name:         cosmic-lamb-jwjwm
    Namespace:
    Labels:       ap.apple.com/node-group=cosmic-lamb
                  karpenter.k8s.aws/instance-category=m
                  karpenter.k8s.aws/instance-cpu=4
                  karpenter.k8s.aws/instance-cpu-manufacturer=intel
                  karpenter.k8s.aws/instance-encryption-in-transit-supported=false
                  karpenter.k8s.aws/instance-family=m5
                  karpenter.k8s.aws/instance-generation=5
                  karpenter.k8s.aws/instance-hypervisor=nitro
                  karpenter.k8s.aws/instance-memory=16384
                  karpenter.k8s.aws/instance-network-bandwidth=1250
                  karpenter.k8s.aws/instance-size=xlarge
                  karpenter.sh/capacity-type=on-demand
                  karpenter.sh/nodepool=cosmic-lamb
                  kubernetes.io/arch=amd64
                  kubernetes.io/os=linux
                  node.kubernetes.io/instance-type=m5.xlarge
                  topology.kubernetes.io/region=us-west-2
                  topology.kubernetes.io/zone=us-west-2a
    Annotations:  karpenter.k8s.aws/ec2nodeclass-hash: 2702429556956267588
                  karpenter.k8s.aws/ec2nodeclass-hash-version: v2
                  karpenter.k8s.aws/tagged: true
                  karpenter.sh/nodepool-hash: 4656253658582728767
                  karpenter.sh/nodepool-hash-version: v2
                  meta.helm.sh/release-name: cosmic-lamb
                  meta.helm.sh/release-namespace: cosmic-lamb
    API Version:  karpenter.sh/v1beta1
    Kind:         NodeClaim
    Metadata:
      Creation Timestamp:  2024-07-10T22:31:25Z
      Finalizers:
        karpenter.sh/termination
      Generate Name:  cosmic-lamb-
      Generation:     1
      Managed Fields:
        API Version:  karpenter.sh/v1beta1
        Fields Type:  FieldsV1
        fieldsV1:
          f:metadata:
            f:annotations:
              .:
              f:karpenter.k8s.aws/ec2nodeclass-hash:
              f:karpenter.k8s.aws/ec2nodeclass-hash-version:
              f:karpenter.k8s.aws/tagged:
              f:karpenter.sh/nodepool-hash:
              f:karpenter.sh/nodepool-hash-version:
              f:kubernetes.io/description:
              f:meta.helm.sh/release-name:
              f:meta.helm.sh/release-namespace:
            f:finalizers:
              .:
              v:"karpenter.sh/termination":
            f:generateName:
            f:labels:
              .:
              f:ap.apple.com/node-group:
              f:karpenter.k8s.aws/instance-category:
              f:karpenter.k8s.aws/instance-cpu:
              f:karpenter.k8s.aws/instance-cpu-manufacturer:
              f:karpenter.k8s.aws/instance-encryption-in-transit-supported:
              f:karpenter.k8s.aws/instance-family:
              f:karpenter.k8s.aws/instance-generation:
              f:karpenter.k8s.aws/instance-hypervisor:
              f:karpenter.k8s.aws/instance-memory:
              f:karpenter.k8s.aws/instance-network-bandwidth:
              f:karpenter.k8s.aws/instance-size:
              f:karpenter.sh/capacity-type:
              f:karpenter.sh/nodepool:
              f:kubernetes.io/arch:
              f:kubernetes.io/os:
              f:node.kubernetes.io/instance-type:
              f:topology.kubernetes.io/region:
              f:topology.kubernetes.io/zone:
            f:ownerReferences:
              .:
              k:{"uid":"2b45b290-f995-4146-ad4c-6a7af5948522"}:
          f:spec:
            .:
            f:nodeClassRef:
              .:
              f:apiVersion:
              f:kind:
              f:name:
            f:requirements:
            f:resources:
              .:
              f:requests:
                .:
                f:cpu:
                f:ephemeral-storage:
                f:memory:
                f:pods:
        Manager:      karpenter
        Operation:    Update
        Time:         2024-07-10T22:32:23Z
        API Version:  karpenter.sh/v1beta1
        Fields Type:  FieldsV1
        fieldsV1:
          f:status:
            .:
            f:allocatable:
              .:
              f:cpu:
              f:ephemeral-storage:
              f:memory:
              f:pods:
              f:vpc.amazonaws.com/pod-eni:
            f:capacity:
              .:
              f:cpu:
              f:ephemeral-storage:
              f:memory:
              f:pods:
              f:vpc.amazonaws.com/pod-eni:
            f:conditions:
            f:imageID:
            f:nodeName:
            f:providerID:
        Manager:      karpenter
        Operation:    Update
        Subresource:  status
        Time:         2024-07-10T22:32:31Z
      Owner References:
        API Version:           karpenter.sh/v1beta1
        Block Owner Deletion:  true
        Kind:                  NodePool
        Name:                  cosmic-lamb
        UID:                   2b45b290-f995-4146-ad4c-6a7af5948522
      Resource Version:        391169015
      UID:                     d698e25a-8f78-443f-900d-b34b7a39c35a
    Spec:
      Node Class Ref:
        API Version:  karpenter.k8s.aws/v1beta1
        Kind:         EC2NodeClass
        Name:         cosmic-lamb
      Requirements:
        Key:       kubernetes.io/arch
        Operator:  In
        Values:
          amd64
        Key:       node.kubernetes.io/instance-type
        Operator:  In
        Values:
          m5.xlarge
        Key:       ap.apple.com/node-group
        Operator:  In
        Values:
          cosmic-lamb
        Key:       karpenter.sh/nodepool
        Operator:  In
        Values:
          cosmic-lamb
        Key:       karpenter.k8s.aws/instance-size
        Operator:  In
        Values:
          xlarge
        Key:       kubernetes.io/os
        Operator:  In
        Values:
          linux
        Key:       topology.kubernetes.io/zone
        Operator:  In
        Values:
          us-west-2a
        Key:       karpenter.k8s.aws/instance-generation
        Operator:  Gt
        Values:
          2
        Key:       karpenter.sh/capacity-type
        Operator:  In
        Values:
          on-demand
        Key:       karpenter.k8s.aws/instance-hypervisor
        Operator:  In
        Values:
          nitro
        Key:       karpenter.k8s.aws/instance-family
        Operator:  In
        Values:
          m5
      Resources:
        Requests:
          Cpu:                  3190m
          Ephemeral - Storage:  48Gi
          Memory:               12880Mi
          Pods:                 54
    Status:
      Allocatable:
        Cpu:                        3920m
        Ephemeral - Storage:        89Gi
        Memory:                     14162Mi
        Pods:                       58
        vpc.amazonaws.com/pod-eni:  18
      Capacity:
        Cpu:                        4
        Ephemeral - Storage:        100Gi
        Memory:                     15155Mi
        Pods:                       58
        vpc.amazonaws.com/pod-eni:  18
      Conditions:
        Last Transition Time:  2024-07-10T22:32:31Z
        Status:                True
        Type:                  Initialized
        Last Transition Time:  2024-07-10T22:31:26Z
        Status:                True
        Type:                  Launched
        Last Transition Time:  2024-07-10T22:32:31Z
        Status:                True
        Type:                  Ready
        Last Transition Time:  2024-07-10T22:32:21Z
        Status:                True
        Type:                  Registered
      Image ID:                ami-002c17c43a2d082b5
      Node Name:               ip-xxx-xx-xxx-xxx.us-west-2.compute.internal
      Provider ID:             aws:///us-west-2a/i-024721b0fc907808a
    Events:
      Type    Reason             Age                 From       Message
      ----    ------             ----                ----       -------
      Normal  DisruptionBlocked  55m                 karpenter  Cannot disrupt NodeClaim: Nominated for a pending pod
      Normal  Unconsolidatable   8m1s (x4 over 54m)  karpenter  Can't replace with a cheaper node

</details>

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/apis/crds/karpenter.k8s.aws_ec2nodeclasses.yaml [hunk 1]
```diff
@@ -758,6 +758,13 @@ spec:
                     - Windows2019
                     - Windows2022
                   type: string
+                amiIncludeDeprecated:
```
Comment (njtran) [inline pkg/apis/crds/karpenter.k8s.aws_ec2nodeclasses.yaml:761]: I don't know if we want to add the API surface here. I think we should be able to do this, and it just works:
1. Prioritize AMIs that are not deprecated, prioritizing the newest AMI.
2. If there are no non-deprecated AMIs, use the newest of the deprecated AMIs.

**Review by njtran — COMMENTED** [review]
One high level comment that I think would change most of the PR. Let me know what you think

File: pkg/apis/crds/karpenter.k8s.aws_ec2nodeclasses.yaml (hunk 1 above)
Comment (shabbskagalwala) [reply pkg/apis/crds/karpenter.k8s.aws_ec2nodeclasses.yaml:761]: @njtran that would mean that the `IncludeDeprecated` field is always set to true on the api call, right? 

By default, if the current account is the AMI owner, all deprecated AMIs appear in the response regardless. If the AMI id specified is a cross account ami that becomes an issue. That AMI is not discoverable at all without this parameter. 

My understanding was that when tags or names are used Karpenter will handle the AMI discovery priortizing the newest AMI always - the issue only comes when pinning using ami ids and if the ami is owned by a different account. 

>Prioritize AMIs that are not deprecated, prioritizing the newest AMI.

This is how it works currently. 
 
>If there are no non-deprecated AMIs, use the newest of the deprecated AMIs.
For this do you think, that if no ami's are returned then try using the `IncludeDeprecated` attribute in the sdk call to describe images?

Comment (njtran) [reply pkg/apis/crds/karpenter.k8s.aws_ec2nodeclasses.yaml:761]: Yeah I think we can just enable it by default in the call, is there a reason not to include it by default? When it comes to priority, Karpenter should currently prioritize the newest AMI regardless if it's tags, names, or ID. We'd just get all the AMIs, both deprecated and not, and filter the results

**Pull request closed without merging**